nucleus/flashlight/framework/src/main/java/org/glassfish/flashlight/impl/client/ProbeProviderClassFileTransformer.java

```diff
@@ -79,141 +78,94 @@
+        updater.updateAndGet(thread -> thread != null && thread.isAlive() ? thread : newUpdater());
     }
 
-    final synchronized void transform() {
-
-        Class providerClass = providerClassRef.get();
+    private final void runUpdate() {
+        faggedForUpdate.set(false);
+        Class<?> providerClass = providerClassRef.get();
         if (providerClass == null) {
             if (Log.getLogger().isLoggable(Level.FINER))
                 Log.finer("provider class was reclaimed, not.transformed", providerClassName);
-            return; // Nothing to do!            
-        }
-            
-        if (enabled)  {
-            if (allProbesTransformed) {
-                if (Log.getLogger().isLoggable(Level.FINER))
-                    Log.finer("all probes already.transformed", providerClass);
-                return; // Nothing to do!
-            }
-            if (Log.getLogger().isLoggable(Level.FINER))
-                Log.finer("some probes need to be.transformed", providerClass);
+            return; // Nothing to do!
         }
-        allProbesTransformed = true;
-        
-        //important!  The transform(...) callback method in this class uses this boolean!
-        enabled = true;
+        if (Log.getLogger().isLoggable(Level.FINER))
+            Log.finer("some probes need to be.transformed", providerClass);
 
         if (instrumentation == null)
             return;
-
         try {
             addTransformer();
             instrumentation.retransformClasses(providerClass);
-        }
-        catch (Exception e) {
+        } catch (Exception e) {
             logger.log(Level.WARNING, RETRANSFORMATION_ERROR, e);
         }
     }
 
-    final synchronized void untransform() {
-        Class providerClass = providerClassRef.get();
-        if (providerClass == null) {
-            if (Log.getLogger().isLoggable(Level.FINER))
-                Log.finer("provider class was reclaimed, not.untransformed", providerClassName);
-            return; // Nothing to do!            
-        }
-        
-        // Reset the flag to indicate the probes are not transformed
-        allProbesTransformed = false;
-        
-        if (!enabled) {
-            if (Log.getLogger().isLoggable(Level.FINER))
-                Log.finer("already.not.transformed", providerClass);
-            return; // Nothing to do!
-        }
-
-        //important!  The transform(...) callback method in this class uses this boolean!
-        enabled = false;
-
-        if (instrumentation == null)
-            return;
-
-        try {
-            instrumentation.retransformClasses(providerClass);
-        }
-        catch (UnmodifiableClassException e) {
-            logger.log(Level.WARNING, RETRANSFORMATION_ERROR, e);
+    private boolean hasEnabledInvokers() {
+        for (FlashlightProbe probe : probes.values()) {
+            if (probe.isEnabled() && probe.getInvokerCount() > 0) {
+                return true;
+            }
```

Comment:
Instead of trying to maintain a "dirty" flag telling us if instrumentation needs to be updated the decision is now made based on the probe state. Anything else would cause us to remove instrumentation when there are active invokers waiting to be called.